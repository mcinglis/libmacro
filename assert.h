
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ASSERT_H
#define LIBMACRO_ASSERT_H


#include <assert.h>

#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON


#ifdef NO_ASSERT
    #define ASSERT( ... )
#else

    // @public
    // Takes a variable number of scalar expressions, and `assert`s each
    // one individually. This provides better assertion error reporting than
    // techniques like `assert( e1 && e2 && ... )`.
    #define ASSERT( ... ) \
        PP_MAP( assert, PP_SEP_SEMICOLON, __VA_ARGS__ )

#endif


#endif


