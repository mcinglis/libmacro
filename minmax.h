
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_MINMAX_H
#define LIBMACRO_MINMAX_H


#include <libpp/foldr.h>        // PP_FOLDR


// @public begin


// Returns the minimum value of the two given scalars.
#define MIN2( X, Y ) \
    (((X)<(Y))?(X):(Y))

// Returns the minimum value of the `n` given scalars.
#define MIN( ... ) \
    PP_FOLDR( MIN2, __VA_ARGS__ )


// Returns the maximum value of the two given scalars.
#define MAX2( X, Y ) \
    (((X)>(Y))?(X):(Y))

// Returns the maximum value of the `n` given scalars.
#define MAX( ... ) \
    PP_FOLDR( MAX2, __VA_ARGS__ )


// @public end


#endif

