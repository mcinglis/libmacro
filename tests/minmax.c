
#include <stdio.h>

#include "../assert.h"          // ASSERT
#include "../minmax.h"          // MIN, MAX


int main( void )
{
    printf( "Testing MIN of two values...\n" );
    ASSERT( MIN( -1, 1 ) == -1,
            MIN( 0, 1 )  == 0,
            MIN( 1, 1 )  == 1,
            MIN( 0, -1 ) == -1,
            MIN( 9182, -129 ) == -129 );

    printf( "Testing MIN of many values...\n" );
    ASSERT( MIN( 0, 1, 2 ) == 0,
            MIN( 45, 87, -1 ) == -1,
            MIN( -123, -456, -789 ) == -789 );

    printf( "Testing MAX of two values...\n" );
    ASSERT( MAX( -1, 1 ) == 1,
            MAX( 0, 1 )  == 1,
            MAX( 1, 1 )  == 1,
            MAX( 0, -1 ) == 0,
            MAX( 9182, -129 ) == 9182 );

    printf( "Testing MAX of many values...\n" );
    ASSERT( MAX( 0, 1, 2 ) == 2,
            MAX( 45, 87, -1 ) == 87,
            MAX( -123, -456, -789 ) == -123 );
}

