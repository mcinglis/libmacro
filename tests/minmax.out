Testing MIN of two values...
Testing MIN of many values...
Testing MAX of two values...
Testing MAX of many values...
